src/Utilities/results.jl

```diff
@@ -0,0 +1,283 @@
+    # TODO do not include constant if primal solution is a ray
+    return evalvariables(vi -> MOI.get(model, MOI.VariablePrimal(), vi), f)
+end
+
+################ Constraint Dual for Variable-wise constraints #################
+#
+# In the primal we have
+#   min a_0' x + b_0
+#       A_i  x + b_i in C_i for all i
+# In the dual we have
+#   max b_0 - sum b_i' y
+#       a_0 - sum A_i* y_i = 0
+#                      y_i in C_i* for all i
+# where A_i* is the adjoint operator of the linear operator A_i. That is, A*
+# is the linear operator such that
+# ⟨A x, y⟩_{C_i} = ⟨x, A* y⟩_Rn
+# where
+# * ⟨., .⟩_Rn is the standard scalar product over Rn: ⟨., .⟩_Rn and
+# * ⟨., .⟩_{C_i} is the scalar product `set_dot` defined for the set C_i
+#
+# Suppose we want to get the constraint variable of a variable-wise constraint:
+#   A_j x in C_j
+# where A_j is zero except on a submatrix which is the identity. We have
+# A_j* y_j = a_0 - sum_(i != j) A_i* y_i
+# Thus to get the dual y_j, we simply have to compute the right-hand side and
+# then invert A_j*. To get the kth element of A_i* y_i we need to compute
+# ⟨e_k, A_i* y_i⟩_Rn = ⟨A_i e_k, y_i⟩_{C_i}. A_i e_k is computed using
+# `variable_coefficient` and then it is combined with the dual y_i with
+# `set_dot`.
+# Once A_j* y_j is obtained, we invert A_j* with `dot_coefficients`.
+
+function variable_coefficient(func::MOI.ScalarAffineFunction{T},
+                              vi::MOI.VariableIndex) where T
+    coef = zero(T)
+    for term in func.terms
+        if term.variable_index == vi
+            coef += term.coefficient
+        end
+    end
+    return coef
+end
+function variable_coefficient(func::MOI.VectorAffineFunction{T},
+                              vi::MOI.VariableIndex) where T
+    coef = zeros(T, MOI.output_dimension(func))
+    for vector_term in func.terms
+        term = vector_term.scalar_term
+        if term.variable_index == vi
+            coef[vector_term.output_index] += term.coefficient
+        end
+    end
+    return coef
+end
+
+function variable_dual(model::MOI.ModelLike,
+                       ::MOI.ConstraintDual,
+                       ci::MOI.ConstraintIndex,
+                       vi::MOI.VariableIndex,
+                       F::Type{<:Union{MOI.SingleVariable,
+                                       MOI.VectorOfVariables}},
+                       S::Type{<:MOI.AbstractSet})
+    for constraint_index in MOI.get(model, MOI.ListOfConstraintIndices{F, S}())
+        if constraint_index != ci
+            func = MOI.get(model, MOI.ConstraintFunction(), constraint_index)
+            if (F == MOI.SingleVariable && func.variable == vi) ||
+               (F == MOI.VectorOfVariables && vi in func.variables)
+               error("Fallback getter for variable constraint dual does not support other variable-wise constraints on the variable.")
+            end
+        end
+    end
+    return 0.0
+end
+
+function variable_dual(::MOI.ModelLike,
```

Comment:
This function has lots of methods that aren't obvious to understand. It needs a docstring or at least a comment.